src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,982 @@
+ * - We cannot handle the case where neighbors have more/fewer
+ *   coefficients. In the case that the neighbor has more coefficients we can
+ *   just ignore the higher coefficients. In the case that the neighbor has
+ *   fewer coefficients we have a few choices that will probably depend on which
+ *   specific HL is used.
+ * - Having different number of collocation points in different directions is
+ *   not supported. However, it is straightforward to handle this case. The
+ *   outer most loop is in the direction with the most collocation points, while
+ *   the inner most loop is over the direction with the fewest collocation
+ *   points. The highest to lowest coefficients can then be limited
+ *   appropriately again.
+ * - h-refinement is not supported, but is straightforward to add since the
+ *   differences in coefficients in Eq.\f$\ref{eq:krivodonova 3d minmod}\f$ will
+ *   just be multiplied by non-unity factors.
+ */
+template <size_t VolumeDim, typename... Tags>
+class Hierarchical<VolumeDim, tmpl::list<Tags...>> {
+ public:
+  /*!
+   * \brief The type of hierarchical limiter.
+   *
+   * Currently only the Krivodonova limiter is implemented.
+   */
+  struct Type {
+    using type = HierarchicalType;
+    static constexpr OptionString help = {"Type of hierarchical limiter"};
+  };
+
+  /*!
+   * \brief The \f$\alpha_i\f$ values in the Krivodonova algorithm.
+   */
+  struct Alphas {
+    using type = std::vector<double>;
+    static constexpr OptionString help = {"The alpha coefficients"};
+    static type default_value() noexcept {
+      return std::vector<double>(
+          Spectral::maximum_number_of_points<Spectral::Basis::Legendre>, 1.0);
+    }
+  };
+  /*!
+   * \brief Turn the limiter off
+   *
+   * This option exists to temporarily disable the limiter for debugging
+   * purposes. For problems where limiting is not needed, the preferred
+   * approach is to not compile the limiter into the executable.
+   */
+  struct DisableForDebugging {
+    using type = bool;
+    static type default_value() noexcept { return false; }
+    static constexpr OptionString help = {"Disable the limiter"};
+  };
+
+  using options = tmpl::list<Type, Alphas, DisableForDebugging>;
+  static constexpr OptionString help = {
+      "A collection of hierarchical slope limiters.\n\n"
+      "Currently only the Krivodonova limiter is implemented.\n\n"
+      "Hierarchical slope limiters work by limiting the highest "
+      "coefficients/derivatives using an aggressive approach such as minmod, "
+      "decreasing in coefficient order until no more limiting is necessary."};
+
+  explicit Hierarchical(
+      HierarchicalType hierarchical_type,
+      std::vector<double> alphas = std::vector<double>(
+          Spectral::maximum_number_of_points<Spectral::Basis::Legendre>, 1.0),
+      bool disable_for_debugging = false, const OptionContext& contex = {});
+
+  Hierarchical() = default;
+  Hierarchical(const Hierarchical&) = default;
```

Comment:
do you need the copy constructor?

Reply:
I did because of a DataBox perfect forwarding bug, I'm adding a commit that fixes that so copy semantics won't be needed.